src/test/java/com/google/sps/tests/UrlRequestTest.java

```diff
@@ -0,0 +1,137 @@
+import org.hamcrest.MatcherAssert;
+import static org.hamcrest.collection.IsIterableContainingInAnyOrder.containsInAnyOrder;
+
+/** */
+@RunWith(PowerMockRunner.class) 
+@PrepareForTest({ URL.class, UrlRequest.class })
+public final class UrlRequestTest {
+
+  private final static String EMPTY_STRING = "";
+  private final static Map<String, String> EMPTY_PARAMS = new HashMap<>();
+  private final static String BASE_PATH = "http://www.testing.com";
+  private final static String URL_PATH = String.format("%s?%s", BASE_PATH, UrlRequest.getParamsString(EMPTY_PARAMS));
+
+  private static HttpURLConnection mockUrlConnection(String urlPath) throws Exception {
+    URL mockedUrl = PowerMockito.mock(URL.class);
+    HttpURLConnection mockedConnection = PowerMockito.mock(HttpURLConnection.class);
+
+    PowerMockito.whenNew(URL.class).withArguments(urlPath).thenReturn(mockedUrl);
+    PowerMockito.when(mockedUrl.openConnection()).thenReturn(mockedConnection);
+    return mockedConnection;
+  }
+
+  @Test
+  public void errorUrlQuery() throws Exception {
+    int responseCode = 400;
+    HttpURLConnection mockedConnection = mockUrlConnection(URL_PATH);
+    PowerMockito.when(mockedConnection.getResponseCode()).thenReturn(responseCode);
+
+    String actual = UrlRequest.urlQuery(BASE_PATH, EMPTY_PARAMS);
+    String expected = String.format("{\"error\":%s}", responseCode);
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void successfulURlQuery() throws Exception {
+    int responseCode = 200;
+    String content = "hello world";
+    InputStream inputStream = new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8.toString()));
+    HttpURLConnection mockedConnection = mockUrlConnection(URL_PATH);
+    PowerMockito.when(mockedConnection.getResponseCode()).thenReturn(responseCode);
+    PowerMockito.when(mockedConnection.getInputStream()).thenReturn(inputStream);
+
+    String actual = UrlRequest.urlQuery(BASE_PATH, EMPTY_PARAMS);
+    String expected = content;
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void formatEmptyParamsMap() {
+    String actual = UrlRequest.getParamsString(EMPTY_PARAMS);
+    String expected = EMPTY_STRING;
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void formatOneParam() {
+    Map<String, String> params = new HashMap<>();
+    params.put("hello", "world");
+    String actual = UrlRequest.getParamsString(params);
+    String expected = "hello=world";
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void formatMultipleParams() {
+    Map<String, String> params = new HashMap<>();
+    params.put("hello", "world");
+    params.put("second", "param2");
+    params.put("third", "param3");
+    String response = UrlRequest.getParamsString(params);
+    List<String> actual = Arrays.asList(response.split("&", 0));
+    List<String> expected = Arrays.asList("hello=world", "second=param2", "third=param3");
```

Comment:
[Nit] When writing tests it is good to have as little logic post processing the actual value as possible.  In this case, why not just let 'actual' be the string itself, instead of splitting on '&' and then set expected to 'hello=world&second=param2&third=param3'

Reply:
That's how it was before, but the order in which the params pair ("key=value") are added to the string is arbitrary.

Reply:
Oh, googe point.  This makes sense then.